apps/src/applab/RecordListener.js

```
@@ -0,0 +1,204 @@
+    this.tableHandlers_[tableName].reset();
+  }
+  this.tableHandlers_ = {};
+};
+
+//////////////////////////////////////////////////
+// TableHandler
+//////////////////////////////////////////////////
+
+/**
+ * This class encapsulates all the logic and state required for listening to
+ * changes to a single table.
+ * @param {string} tableName
+ * @param {function} callback
+ * @constructor
+ */
+var TableHandler = function (tableName, callback) {
+  /**
+   * @type {{string}}
+   * @private
+   */
+  this.tableName_ = tableName;
+
+  /**
+   *  @type {function}
+   *  @private
+   */
+  this.callback_ = callback;
+
+  /**
+   * @typedef {Object.<number, string>} IdToJsonMap Map from record id
+   * to a JSON-stringified copy of the corresponding record.
+   */
+
+  /**
+   * IdToJsonMap representing the contents of the table.
+   * We use this to efficiently test whether a record's contents has changed.
+   * @type {IdToJsonMap}
+   * @private
+   */
+  this.idToJsonMap_ = {};
+
+  /**
+   * Timeout id of the pending call to window.setTimeout.
+   * @type {number}
+   * @private
+   */
+  this.timeoutId_ = null;
+
+  this.scheduleNextFetch_();
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.scheduleNextFetch_ = function () {
+  this.timeoutId_ = setTimeout(function recordListener() {
+    this.fetchNewRecords_();
+  }.bind(this), RECORD_INTERVAL);
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.fetchNewRecords_ = function () {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = this.reportRecords_.bind(this, req);
+  var url = '/v3/shared-tables/' + Applab.channelId + '/' + this.tableName_;
+  req.open('GET', url, true);
+  req.send();
+};
+
+/**
+ * @param {XMLHttpRequest} req
+ * @private
+ */
+TableHandler.prototype.reportRecords_ = function (req) {
+  var done = XMLHttpRequest.DONE || 4;
+  if (req.readyState !== done) {
+    return;
+  }
+
+  if (req.status < 200 || req.status >= 300) {
+    // ignore errors
```

I assume we don't want to ignore errors here forever.  Worth a TODO to call out a need for monitoring/logging or user-notification of inability to contact the data API?